### DISCUSSION TITLE
Optimised repainting / use of layers

### DISCUSSION THREAD

**Discussion by Willicious** [graphics32/graphics32#310] [General] [answered]
Hi Anders,

This is from a message on Discord quoting you, regarding scene repainting in NeoLemmix:

> I had read that there were problems with the UI performance ... For some reason the UI is using TImage32 controls but isn't using any of the features provided by TImage32 (such as layers). Without layers, TPaintBox32 would have been a better choice.

>Anyway, when I tried to change things a bit, so each little pixel change didn't lead to a full scene repaint, I discovered that the Graphics32's optimized paint scheme was broken for TImage32. I've quick-fixed it locally and that has improved things a lot.
The main game screen still does a full scene repaint due to the way it's composed but I'm sure that can be fixed (in NeoLemmix) at some point. The best thing would be to utilize the layer system for sprites.

Please note that, whilst I can't completely speak for Namida (NeoLemmix dev), he's made it very clear numerous times that he doesn't intend to do any further NL updates than the ones already planned more than a year ago.

However, I would be very interested in pursuing the potential improvements in SuperLemmix, and I may even back-port any such improvements to NeoLemmix at some point (the future of NL development is currently uncertain, but it's looking likely that Forum users would like its development to continue beyond Namida's final version).

Please don't be put off by my relative inexperience - I have learned a lot over the past year and have implemented some fairly complex features that a number of Forum users didn't even think would be possible. With a bit of guidance, I'm sure the improvements you've suggested can be made, and I'm happy for you to use SuperLemmix to test out any other GR32 bugs/improvements.

I've created a new SuperLemmixPlayer GitHub repo and will push changes to this properly (rather than using the copy-paste method), and I've added you as a collaborator. The old one will be deleted shortly.

I hope this meets with your agreement, the graphics rendering performance of NL/SLX is something I'm particularly interested in making improvements to so I look forward to this being a fruitful collaboration for both of us!

The NL/SLX codebase does in fact use rendering layers (see the top of LemRenderHelpers.pas for a list of the layers used), but perhaps these aren't what you meant?

Best regards,

Will

**Comment by andersmelander** [comment, marked as answer]
I'm afraid you have misunderstood my involvement in NeoLemmix; Beyond my PR there isn't any.
The only reason why I made the NeoLemmix pull request was to pave the way for @CuriousKit 's plans to port NeoLemmix to run on a toaster via FreePascal. See #294.

With regards to layers, Lemmix doesn't use Graphics32 layers at all. Instead it has it's own layer system that renders onto TBitmap32. This is fine, but it means that it doesn't benefit from Graphics32's optimized repaint, which in turn means that it does a full scene repaint every time something changes.
Changing this is no small task. As an experiment I tried modifying the main menu screen so it uses layers for the logo, buttons and animated scroller. This is a much more isolated task and it only took a few hours. The worst part was trying to make sense of the code that applies "glow" to the hot button. Anyway, I abandoned the effort as I didn't really have time to complete it.
As you can see in the screenshot below there are some aliasing issue with the scaled button image and the glow doesn't look as nice as the original, to put it mildly.
![billede](https://github.com/graphics32/graphics32/assets/18141921/5e271396-8053-4068-bbd3-0c90987d9ddd)

**Comment by CuriousKit** [comment]
Sorry for being quiet these past few weeks.  Porting Graphics32 is still on the table, but I've been busy with contract work and a few things with the Free Pascal Compiler.  Come later this week though and we'll see if there are a few things I can get polished up.

**Reply by andersmelander** [reply]
No problem. I'm busy too. I just procrastinate a lot :-)

**Comment by Willicious** [comment]
No worries, just getting the project updated has been a good step in the right direction, and you've helped me to understand better how to work with external commits/branches, which is something that needed to happen.

Anyways, I see that you've accepted the collab invite 😊 please do feel free to have a play around with the codebase anytime you feel like it. I might pop the odd question your way, you've been super helpful so far.
